mantidproject/mantid
# Fitting histogram data

Made necessary changes to fit functions, function domains and Fit algorithm to be able to fit 1D functions to histograms in a "histogram aware" way. It means that in the least squares the calculated values are obtained by integrated the function within each bin. This way produces more accurate results for large bins than when using values at the bin centres.

**To test:**
1. Simple test
   - Run this script
   
   ```
   x = [-1, -0.8,-0.6,-0.4,-0.2,0,0.2,0.4,0.6,0.8,1]
   y = [0.0302241,0.04333437713434279,0.06408122585358178,0.09365777093187931,0.1211189415908434,0.1211189415908435,0.0936577709318793,0.06408122585358172,0.04333437713434274,0.03022406683891948]
   e = [0.01] * len(y)
   ws = CreateWorkspace(x, y, e)
   fun = 'name=Lorentzian,FWHM=0.9'
   Fit(fun, ws, Normalise=True, Output='normal')
   Fit(fun, ws, EvaluationType="Histogram", Output='histo')
   ```
   - Check that parameters of the "histo" fit are closer to the correct values which are `Amplitude=1` and `FWHM=1`.
2. More complex test.
   - Run the script:
   
   ```
   import math
   import numpy
   
   # time range for fit
   xdata=WorkspaceFactory.create("Workspace2D",XLength=401,YLength=400,NVectors=1)
   xdata.dataX(0)[:]=numpy.linspace(-10.0,10.0,401)
   xdata.dataY(0)[:]=numpy.random.normal(loc=0.0,scale=0.01,size=400)
   xdata.dataE(0)[:]=0.1
   AnalysisDataService.addOrReplace("xdata",xdata)
   
   BuiltIn="name=Lorentzian"
   
   # evaluate the function and add noise
   ytheory=EvaluateFunction(Function=BuiltIn+",Amplitude=20.0,PeakCentre=0.0,FWHM=1.0",InputWorkspace=xdata,OutputWorkspace="ytheory",EvaluationType='Histogram')
   yexper=Plus(LHSWorkspace=xdata,RHSWorkspace=ytheory,OutputWorkspace="DynamicRange")
   Guesses=(20,0.1,1.1)
   # adjust the X axis and repeat the fits. Print the parameters adjusted back to a line centred at 0.
   Results=WorkspaceFactory.createTable()
   Results.addColumn("int","Bunching",plottype=1)
   Results.addColumn("float","stdAmp",plottype=2)
   Results.addColumn("float","stdAmpErr",plottype=5)
   Results.addColumn("float","stdCen",plottype=2)
   Results.addColumn("float","stdCenErr",plottype=5)
   Results.addColumn("float","stdFWHM",plottype=2)
   Results.addColumn("float","stdFWHMErr",plottype=5)
   Results.addColumn("float","stdCost",plottype=2)
   Results.addColumn("float","awareAmp",plottype=2)
   Results.addColumn("float","awareAmpErr",plottype=5)
   Results.addColumn("float","awareCen",plottype=2)
   Results.addColumn("float","awareCenErr",plottype=5)
   Results.addColumn("float","awareFWHM",plottype=2)
   Results.addColumn("float","awareFWHMErr",plottype=5)
   Results.addColumn("float","awareCost",plottype=2)
   for bun in range(1,31):
   row=[0.0]*15
   row[0]=bun
   ws=Rebunch(InputWorkspace=yexper,NBunch=bun)
   for i,histo in enumerate(('CentrePoint','Histogram')):
       kw = {'Normalise': True} if i == 0 else {}
       ffr=Fit(Function="name=Lorentzian,Amplitude="+str(Guesses[0])+",PeakCentre="+str(Guesses[1])+",FWHM="+str(Guesses[2]),
               InputWorkspace=ws,WorkspaceIndex=1,EvaluationType=histo,Output="HistoFit", **kw)
       params=ffr[3].column(1)
       errors=ffr[3].column(2)
       row[i*7+1:i*7+8]=[params[0],errors[0],params[1],errors[1],params[2],errors[2],params[3]]
       print bun,params[0],errors[0],params[1],errors[1],params[2],errors[2],params[3]
   Results.addRow(row)
   AnalysisDataService.addOrReplace("HistoAwareResults",Results)
   ```
   - This script generates a Lorentzian peak with some random noise then rebunches it in a loop gradually increasing the bin size. For each bin size the data gets fitted twice: first the usual way, then the "histogram aware" way.
   - The results of the fits are in the "HistoAwareResults" table. Parameters from the normal fits are prefixed with "std" and those obtained the new way are prefixed with "aware".
   - Plot parameter values from this table and check that the "aware" kind has smaller variations.

Fixes #16796.

Release notes added to `docs/source/release/v3.8.0/framework.rst`.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
